### DISCUSSION TITLE
Data manipulations

### DISCUSSION THREAD

**Discussion by Mehdi-YC** [arturo-lang/arturo#408] [Q&A]
Hello , 
first let me say that i can see this scripting oang going places,and i want to ask few questions about it

how does it compare to existing scripting langs in terms of memory & speed , example ruby , python and node

2 nd , do we have any library that panipulates html files ( like bs4)

csv and json too ?

have a good day

**Comment by drkameleon** [comment]
Hi!

> first let me say that i can see this scripting oang going places

Thanks a lot for your kind words! Really appreciate it. I'm doing my best so... every feedback counts. (Btw, I've had a look into your own proglang-related repos and I have to say... good job as well! 😉  I guess once one catch the proglang bug, it's difficult to let go haha. With me, it's been probably the past 25 years... and counting)

> how does it compare to existing scripting langs in terms of memory & speed , example ruby , python and node

The interpreter is actually a stack-based bytecode VM (I can explain more about the internals, if you are interested). Now, in terms of performance: I've implemented the exact same interpreter at least 3-4 times, and even in different languages (D, then C, then Nim, then back in C, and again in Nim), always being too obsessed about its performance, benchmarking everything, and micro-optimizing everything too. The result? I had some thing ultra-fast but not complete at all. So, I decided to switch strategies and go for something complete first, and optimize later on. That's exactly the stage we are at. The language *is* fully functional, but I guess there are a lot of optimization to be done. That being said, in practical, everyday-use, terms, I don't see much of a difference between Ruby (the language I mostly use for scripting) and Arturo. And... let me add that the language is already in use in a production environment - [DELE Ahora](https://deleahora.com/), a Spanish-learners website I've created, part of the backend IS in Arturo...)

> 2 nd , do we have any library that panipulates html files ( like bs4)

I had started "modules" for both (https://github.com/arturo-lang/art-html-module & https://github.com/arturo-lang/art-bootstrap-module) but all things going on, and until I make the module system *really* work, and do so properly (which I'm working on), I decided to put them a bit aside. But I *will* get back to it.

What exactly would you expect from an HTML module?

Right now, the main side-project, entirely written in Arturo (and obviously plain HTML/JS/etc for the frontend is Grafito: https://github.com/arturo-lang/grafito, a Graph database system with an SQLite backend, written in Arturo, and even making using of Arturo's lightweight syntax for queries):

```red
person [ 
    isFrom: country "United States"  
    directed: movie [ 
        title:-> contains: "a" 
        year:-> under: 2000
    ]
]
```

The README is not up-to-date. Even the UI doesn't look like that but, hopefully, much better. You could have a look at a sneak peek I've published from the last PR I've merged (but not documented yet 😄 ): https://discord.com/channels/765519132186640445/831551548605595698/935881791905103932

> csv and json too ?

CSV/JSON already come included in Arturo.

Mainly, the core, rather "hidden", functions are [`read`](https://arturo-lang.io/documentation/library/files/read/) and [`write`](https://arturo-lang.io/documentation/library/files/write/). You may think they just work with files (and, yes, after all they are in the Files module - but there had to be put somewhere!), but actually they work with anything: strings, files, url content, anything.

And you can handle it as you please.

**E.g.**

<img width="779" alt="Screenshot 2022-01-29 at 10 04 13" src="https://user-images.githubusercontent.com/1265028/151654990-3b4e5d97-c0c5-4be7-9d3c-2223e22dfc7c.png">

Let me know your thoughts about the whole thing. I'm definitely interested! 😉 

Cheers from Spain! :)

**Comment by Mehdi-YC** [comment]
Thank you for this details answer , i guess that languages took over your life too haha ! 
well from an HTML module , not the html itself , but something to parse it , or more like query selector like the JS one to extract data easily 
im interested at making one and help you in this project once i get a little bit of time , right now im trying t setup everything and have time to make videos and blogs too , and it will be interesting to do some about this scripting lang.
